subsys/shell/backends/shell_rtt.c

```diff
@@ -78,29 +85,82 @@ static int enable(const struct shell_transport *transport, bool blocking)
 	struct shell_rtt *sh_rtt = (struct shell_rtt *)transport->ctx;
 
 	if (blocking) {
-		rtt_blocking = true;
 		k_timer_stop(&sh_rtt->timer);
 	}
 
 	return 0;
 }
 
+static inline bool is_panic_mode(void)
+{
+	return panic_mode;
+}
+
+static inline bool is_sync_mode(void)
+{
+	return (IS_ENABLED(CONFIG_LOG_MODE_IMMEDIATE) && IS_ENABLED(CONFIG_SHELL_LOG_BACKEND)) ||
+		is_panic_mode();
+}
+
+static void on_failed_write(int retry_cnt)
+{
+	if (retry_cnt == 0) {
+		host_present = false;
+	} else if (is_sync_mode()) {
+		k_busy_wait(USEC_PER_MSEC *
+				CONFIG_SHELL_BACKEND_RTT_RETRY_DELAY_MS);
+	} else {
+		k_msleep(CONFIG_SHELL_BACKEND_RTT_RETRY_DELAY_MS);
+	}
+}
+
+static void on_write(int retry_cnt)
+{
+	host_present = true;
+	if (is_panic_mode()) {
+		/* In panic mode block on each write until host reads it. This
+		 * way it is ensured that if system resets all messages are read
+		 * by the host. While pending on data being read by the host we
+		 * must also detect situation where host is disconnected.
+		 */
+		while (SEGGER_RTT_HasDataUp(CONFIG_SHELL_BACKEND_RTT_BUFFER) &&
+			host_present) {
+			on_failed_write(retry_cnt--);
+		}
+	}
+
+}
+
 static int write(const struct shell_transport *transport,
 		 const void *data, size_t length, size_t *cnt)
 {
 	struct shell_rtt *sh_rtt = (struct shell_rtt *)transport->ctx;
-	const uint8_t *data8 = (const uint8_t *)data;
+	static bool host_present;
```

Comment:
I am a bit concerned about name shadowing. You have already created host_present as a global variable. Was this intentional?"

Reply:
good finding. That was not intentional. Removed.